File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<int> queue1, vector<int> queue2) {
    int answer = -2;
    // 원형 큐를 생각할까?
    // 전체 합을 먼저 알아야해
    // 전체 합을 세어주는 과정에서 뭔가 알 수 없을까
    long long sum = 0, sum_now = 0;
    vector<int> c_que;
    for (auto q1 : queue1) { c_que.emplace_back(q1); sum += q1; sum_now += q1; }
    for (auto q2 : queue2) { c_que.emplace_back(q2); sum += q2; }
    if (sum % 2 == 1) return -1;
    int p1 = 0, p2 = queue1.size(), cnt = 0;
    while (1) {
        if (sum_now == sum - sum_now) break;
        if (p1 == c_que.size()) p1 = 0;
        if (p2 == c_que.size()) p2 = 0;

        if (sum_now < sum - sum_now) {
            sum_now += c_que[p2];
            p2++;
        } else {
            sum_now -= c_que[p1];
            p1++;
        }
        cnt++;
        if (cnt > c_que.size()*2) {
            cnt = -1;
            break;
        }
    }

    answer = cnt;
    return answer;
}

File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        int answer = -2;
        int[] que = new int[queue1.length + queue2.length];
        long sum = 0, sumNow = 0;
        
        int idx = 0;
        for (int num : queue1) {
            que[idx++] = num;
            sum += num;
            sumNow += num;
        }
        for (int num : queue2) {
            que[idx++] = num;
            sum += num;
        }
        if (sum%2==1) return -1;
        
        int start = 0, end = queue1.length - 1;
        answer = 0;
        
        while (answer < que.length * 2) {
            if (sumNow == sum / 2) {
                return answer;
            } else if (sumNow < sum/2) {
                if (end < que.length - 1) {
                    sumNow += que[++end];
                } else {
                    end = 0;
                    sumNow += que[end];
                }
                answer++;
            } else {
                if (start < que.length - 1) {
                    sumNow -= que[start++];
                } else {
                    start = 0;
                    sumNow -= que[start];
                }
                answer++;
            }
        }
        
        return -1;
    }
}